### PR TITLE
Adding AMatrix library include files to external libraries

### PR DESCRIPTION
**Pull request by pooyan-dadvand**
I have just uploaded the minimum header files required to compile Kratos with AMatrix. In future we may add the rest of the library including the tests and benchmarks.

Adding the difinition `-DAMATRIX_DIR=/path/to/kratos/external_libraries/a_matrix` to configure make CMake compile Kratos with AMatrix

### CODE REVIEW

**Review by jcotela — APPROVED** [review]
Approving to avoid blocking the migration effort.

**Comment by josep-m-carbonell** [comment]
I think the correct addition is: 
-DAMATRIX_DIR="/path/to/kratos/external_libraries/a_matrix" 
(without the  /include folder)

**Comment by josep-m-carbonell** [comment]
Compilation problems with operators: range, project, matrix_row, row, sum ... 
ie.:
ip_w /= sum( ip_w );
project( T, range(3, 6), range(3, 6) ) = project( T, range(0, 3), range(0, 3) );
noalias(EigenVector) = matrix_row<const MatrixType>(rVariables.Strain.Eigen.Vectors,i);
is there any translation of them  for Amatrix?

File: external_libraries/a_matrix/include/matrix_expression.h [hunk 1]
```diff
@@ -0,0 +1,755 @@
...
+    using data_type = typename TExpressionType::data_type;
+    MatrixRow() = delete;
+
+    MatrixRow(TExpressionType& Original, std::size_t RowIndex)
+        : _original_expression(Original), _row_index(RowIndex) {}
+
+    template <typename TOtherExpressionType, std::size_t TCategory>
+    MatrixRow& operator=(
+        MatrixExpression<TOtherExpressionType, TCategory> const& Other) {
+        auto& other_expression = Other.expression();
```
Comment (josep-m-carbonell) [inline external_libraries/a_matrix/include/matrix_expression.h:84]: unused variable **other_expression**

**Comment by josep-m-carbonell** [comment]
Also problems in this type of template method definitions:
```
template<class TMatrix3x3>	
inline void ToRotationMatrix(TMatrix3x3& R)const
{
        if( R.size1()!=3 || R.size2()!=3 )
          R.resize(3,3,false);
       ....
```
error:
```
amatrix_interface.h:181:36: error: too many arguments to function
      call, expected single argument 'NewSize', have 2 arguments
        base_type::resize(NewSize1,NewSize2);
        ~~~~~~~~~~~~~~~~~          ^~~~~~~~
Kratos/kratos/utilities/quaternion.h:224:15: note: in instantiation of member function
      'Kratos::Internals::Matrix<double, 3, 3>::resize' requested here
                          R.resize(3,3,false);
```

**Comment by josep-m-carbonell** [comment]
helper_classes_for_constraint_builder.h  does not compile:

```
In file included from add_strategies_to_python.cpp:57:
In file included from /Kratos/kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h:30:
/Kratos/kratos/utilities/helper_classes_for_constraint_builder.h:40:9: error: use of class
      template 'Matrix' requires template arguments
typedef Matrix MatrixType;
        ^
```
----------------------------------------------
feast_solver.h  does not compile:

```
/Kratos/applications/ExternalSolversApplication/external_includes/feast_solver.h:298:50: error: 
      member reference base type 'double *' is not a structure or union
                    (double *)rEigenvalues.data().begin(),
                              ~~~~~~~~~~~~~~~~~~~^~~~~~
/Kratos/applications/ExternalSolversApplication/external_includes/feast_solver.h:347:51: error: 
      no matching function for call to 'prod'
                        noalias(column(work,k)) = prod(rMassMatrix,row(rEigenvectors,k));
```
---------------------------------------------
also problems with pastix:

```
/Kratos/applications/ExternalSolversApplication/external_includes/pastix_complex_solver.h:167:3: error: 
      call to 'noalias' is ambiguous
                noalias(rX) = rB;
```

**Comment by pooyan-dadvand** [comment]
@josep-m-carbonell Thanks for checking it and your comments!

And answering your questions:

* There is no sum yet for kratos matrix. I should add one. 
* instead of `project` you may use `AMatrix::SubMatrix<MatrixType>a_sub_matrix(original_matrix, sub_index1, sub_index2, sub_size1, sub_size2);` as you may appreciate the argument are different. So I would suggest you to use it with `KRATOS_USE_AMATRIX` in order to maintain the compatibility to ublas
*  instead of using the `matrix_row(rVariables.Strain.Eigen.Vectors,i)` you may use the `row((rVariables.Strain.Eigen.Vectors,i)` defined in kratos.

* you should pass false to your resize methods to avoid the following error
```
amatrix_interface.h:181:36: error: too many arguments to function
      call, expected single argument 'NewSize', have 2 arguments
        base_type::resize(NewSize1,NewSize2);
        ~~~~~~~~~~~~~~~~~          ^~~~~~~~
Kratos/kratos/utilities/quaternion.h:224:15: note: in instantiation of member function
      'Kratos::Internals::Matrix<double, 3, 3>::resize' requested here
                          R.resize(3,3,false);
```


I have added a wiki [page](https://github.com/KratosMultiphysics/Kratos/wiki/Porting-to-AMatrix) in order to put all these changes with previous ones together. Please check the change list there for other errors you got. It would be great to add an error and solution section to it.

**Comment by jcotela** [comment]
@pooyan-dadvand The wiki page mentions using ScalarMatrix instead of Matrix to initialize a matrix to a scalar value. Is there a python equivalent of that? several python tests in core are failing because they try to initialize a matrix (or vector) to some value.

**Comment by pooyan-dadvand** [comment]
Thanks for bug report. 

In AMatrix this constructor is omitted as it gives ambiguity error for vector version and matrix one. But in python they are exported with different names so we may trick the init for that. I will check it.

**Comment by RiccardoRossi** [comment]
Important problem: we should mimic the behaviour of ublas clear... or remove it for good

**Comment by pooyan-dadvand** [comment]
The `clear` method is added with the same behavior as ublas

**Comment by josep-m-carbonell** [comment]
Hi @pooyan-dadvand, 

> There is no sum yet for kratos matrix. I should add one.

Ok we must wait to have one to change to Amatrix

> instead of project you may use AMatrix::SubMatrix<MatrixType>a_sub_matrix(original_matrix, sub_index1, sub_index2, sub_size1, sub_size2); as you may appreciate the argument are different. So I would suggest you to use it with KRATOS_USE_AMATRIX in order to maintain the compatibility to ublas

I find this method a little bit confusing, can this method be similar to the ublas one ?
the operation to be translated is this one:
project( T, range(3, 6), range(3, 6) ) = project( T, range(0, 3), range(0, 3) );

> instead of using the matrix_row(rVariables.Strain.Eigen.Vectors,i) you may use the row((rVariables.Strain.Eigen.Vectors,i) defined in kratos.

Ok. However we have the expressions:
  typedef matrix_row< const MatrixType > ConstMatrixRowType;

> you should pass false to your resize methods to avoid the following error
> 
> amatrix_interface.h:181:36: error: too many arguments to function
>       call, expected single argument 'NewSize', have 2 arguments
>         base_type::resize(NewSize1,NewSize2);
>         ~~~~~~~~~~~~~~~~~          ^~~~~~~~
> Kratos/kratos/utilities/quaternion.h:224:15: note: in instantiation of member function
>       'Kratos::Internals::Matrix<double, 3, 3>::resize' requested here
>                           R.resize(3,3,false);
> I have added a wiki page in order to put all these changes with previous ones together. Please check the change list there for other errors you got. It would be great to add an error and solution section to it.

The problem here is that the method is a template of MatrixType, but sometimes a BoundedMatrix is passed instead of a Matrix as a template argument. I guess that ublas was able to handle this (ignoring to resize a bounded matrix).... I will try to not pass a BoundedMatrix.

Massimo was using this type of operators project, range, sum, matrix_row, so, now we have them in the shell utilities, in the SolidMechanics and in the StructuralMechanics applications.

**Comment by pooyan-dadvand** [comment]
Sorry for late reply, some immediate deadline...

I'm implementing the sum in the core/add-sum-to-matrix branch (compiling...)

> I find this method a little bit confusing, can this method be similar to the ublas one ?
> the operation to be translated is this one:
> project( T, range(3, 6), range(3, 6) ) = project( T, range(0, 3), range(0, 3) );

I understand that it may be confusing coming from ublas but the interface is similar to Eigen. Let me see if I can make a project like function for that. What is an overkill here is the use of range which I would avoid 

> Ok. However we have the expressions:
> typedef matrix_row< const MatrixType > ConstMatrixRowType;

row function also has two versions for const and non const matrices

For the resize problem I see your point and admit that it is a problem. I have chosen to not have a resize to be more restrict with stack types. ublas resizes the stack arrays and for that reason they are slower than Eigen and AMatrix ones. In array_1d I implement a resize which does nothing but I found it misleading during years. So I'm open to new suggestions. Sometimes too generic is also too ambiguos...

**Comment by philbucher** [comment]
I also have to wait with porting StructuralMechanics until this discussion is figured out
Please add me in upcoming issues / PRs related to this

**Comment by RiccardoRossi** [comment]
Regarding the resize, can we mimic the behaviour of old array_1d and eventually remove it at a later stage?

**Comment by pooyan-dadvand** [comment]
We can do this but I'm not sure if it is the best option as someone can think that it is really resizing the vector as ublas does. Actually I found (very few) erroneous resize of array_1d in my cleanup. 

If there are several template functions with this problem I would go for a dummy resize solution. If not IMO we should drop the generality and change those function.

**Pull request merged**